crates/rattler_solve/src/resolvo/mod.rs

```diff
@@ -708,23 +897,47 @@ impl super::SolverImpl for Solver {
     }
 }
 
-fn parse_match_spec<'a>(
-    pool: &Pool<SolverMatchSpec<'a>>,
-    spec_str: &'a str,
-    parse_match_spec_cache: &mut HashMap<&'a str, VersionSetId>,
-) -> Result<VersionSetId, ParseMatchSpecError> {
-    if let Some(spec_id) = parse_match_spec_cache.get(spec_str) {
-        Ok(*spec_id)
+fn parse_match_spec(
+    pool: &Pool<SolverMatchSpec<'_>, NameType>,
+    spec_str: &str,
+    parse_match_spec_cache: &mut HashMap<String, VersionSetId>,
+) -> Result<Vec<VersionSetId>, ParseMatchSpecError> {
+    let match_spec = MatchSpec::from_str(spec_str, ParseStrictness::Lenient)?;
+    let (name, spec) = match_spec.into_nameless();
+
+    let mut version_set_ids = vec![];
+
+    if let Some(ref features) = spec.optional_features {
+        for feature in features {
+            let mut spec_with_feature = spec.clone();
+            spec_with_feature.optional_features = Some(vec![feature.to_string()]);
+
+            let name_with_feature = NameType::WithFeature(
+                name.as_ref()
+                    .expect("Packages with no name are not supported")
+                    .as_normalized()
+                    .to_owned(),
+                feature.to_string(),
+            );
+            let dependency_name = pool.intern_package_name(name_with_feature);
+            let version_set_id = pool.intern_version_set(dependency_name, spec_with_feature.into());
+            version_set_ids.push(version_set_id);
+        }
     } else {
-        let match_spec = MatchSpec::from_str(spec_str, ParseStrictness::Lenient)?;
-        let (name, spec) = match_spec.into_nameless();
+        if let Some(spec_id) = parse_match_spec_cache.get(spec_str) {
+            return Ok(vec![*spec_id]);
+        }
+
         let dependency_name = pool.intern_package_name(
             name.as_ref()
-                .expect("match specs without names are not supported")
+                .expect("Packages with no name are not supported")
                 .as_normalized(),
         );
         let version_set_id = pool.intern_version_set(dependency_name, spec.into());
```

Comment:
Because of the `From` impl this will also insert a set a feature for the SolverMatchSpec!

I would instead change this function such that you create a `SolverMatchSpec` for the NamelessMatchSpec. Without a feature. Then clone that `SolverMatchSpec` for every feature and explicitly set the `feature` field.